pkg/gardenlet/controller/vpaevictionrequirements/reconciler.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2024 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vpaevictionrequirements
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	vpaautoscalingv1 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/gardener/gardener/pkg/utils/timewindow"
+)
+
+var upscaleOnlyRequirement = []*vpaautoscalingv1.EvictionRequirement{{
+	Resources:         []corev1.ResourceName{corev1.ResourceMemory, corev1.ResourceCPU},
+	ChangeRequirement: vpaautoscalingv1.TargetHigherThanRequests,
+}}
+
+// Reconciler implements the reconciliation logic for adding/removing EvictionRequirements to VPA objects.
+type Reconciler struct {
+	Config     config.VPAEvictionRequirementsControllerConfiguration
+	SeedClient client.Client
+	Clock      clock.Clock
+}
+
+// Reconcile implements the reconciliation logic for adding/removing EvictionRequirements to VPA objects.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+	ctx, cancel := controllerutils.GetMainReconciliationContext(ctx, controllerutils.DefaultReconciliationTimeout)
+	defer cancel()
+
+	vpa := &vpaautoscalingv1.VerticalPodAutoscaler{}
+	if err := r.SeedClient.Get(ctx, request.NamespacedName, vpa); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	log.Info("Reconciling")
+	if !metav1.HasLabel(vpa.ObjectMeta, constants.LabelVPAEvictionRequirementDownscaleRestriction) {
```

Comment:
Does `eviction-requirements.autoscaling.gardener.cloud/downscale-restriction` really needs to be a label? I understand that `autoscaling.gardener.cloud/eviction-requirements` has to be a label, because the controller is watching for objects having this label.
If there is no strict requirement, we could make it annotation.
Or do we want the VPA for some reason to be selectable by their `downscale-restriction`?

Reply:
No strong feelings, I'm happy to also make this into a label.

Reply:
done.